packages/block-editor/src/components/block-list/index.js

```diff
@@ -202,122 +169,105 @@ class BlockList extends Component {
 
 		// If the final selection doesn't leave the block, there is no multi
 		// selection.
-		if ( this.startClientId === clientId ) {
-			this.props.onStopMultiSelect();
+		if ( selectedBlockClientId === clientId ) {
+			stopMultiSelect();
 			return;
 		}
 
-		this.props.onMultiSelect( this.startClientId, clientId );
-		this.props.onStopMultiSelect();
-	}
+		multiSelect( selectedBlockClientId, clientId );
+		stopMultiSelect();
+	};
 
-	render() {
-		const {
-			className,
-			blockClientIds,
-			rootClientId,
-			__experimentalMoverDirection: moverDirection = 'vertical',
-			isDraggable,
-			selectedBlockClientId,
-			multiSelectedBlockClientIds,
-			hasMultiSelection,
-			renderAppender,
-			enableAnimation,
-			isMultiSelecting,
-		} = this.props;
-
-		return (
-			<div
-				ref={ this.ref }
-				className={ classnames(
-					'editor-block-list__layout block-editor-block-list__layout',
-					className
-				) }
-			>
-				{ blockClientIds.map( ( clientId, index ) => {
-					const isBlockInSelection = hasMultiSelection ?
-						multiSelectedBlockClientIds.includes( clientId ) :
-						selectedBlockClientId === clientId;
-
-					return (
-						<BlockAsyncModeProvider
-							key={ 'block-' + clientId }
-							clientId={ clientId }
-							isBlockInSelection={ isBlockInSelection }
-						>
-							<BlockListBlock
-								rootClientId={ rootClientId }
-								clientId={ clientId }
-								onSelectionStart={ this.onSelectionStart }
-								isDraggable={ isDraggable }
-								moverDirection={ moverDirection }
-								isMultiSelecting={ isMultiSelecting }
-								// This prop is explicitely computed and passed down
-								// to avoid being impacted by the async mode
-								// otherwise there might be a small delay to trigger the animation.
-								animateOnChange={ index }
-								enableAnimation={ enableAnimation }
-							/>
-						</BlockAsyncModeProvider>
-					);
-				} ) }
-
-				<BlockListAppender
-					rootClientId={ rootClientId }
-					renderAppender={ renderAppender }
-				/>
-			</div>
-		);
-	}
-}
+	/**
+	 * Handles a mouseup event to end the current mouse multi-selection.
+	 */
+	const onSelectionEnd = () => {
```

Comment:
Would it make sense to use [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback) for this and `onSelectionStart`?

Reply:
I'm not sure if would make a noticeable difference.

Reply:
Would love some more opinions for @WordPress/gutenberg-core. :)

Reply:
In general, I tend to think it's not necessary unless we measure performance issues.

That said, I'm pretty sure it's important here to avoid new props for the BlockListBlock component on each render because that's the central component of Gutenberg and the ones that can be rendered a lot of times as the post grows. The difference may not be noticeable I think with the Async mode on but I believe that if you disable the async mode, this could have a big impact.

Reply:
Done